### PR TITLE
solving HW jv-spring-security-part-2

### PR DESCRIPTION
**Pull request by oleh-mazurkov**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -126,10 +126,10 @@
...
                 <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-compiler-plugin</artifactId>
+                <artifactId>maven-war-plugin</artifactId>
+                <version>3.3.1</version>
                 <configuration>
-                    <source>${jdk.version}</source>
-                    <target>${jdk.version}</target>
+                    <failOnMissingWebXml>false</failOnMissingWebXml>
                 </configuration>
             </plugin>
```
Comment (KseniiaMakarova) [inline pom.xml:134]: why have you changed this? you already have added this plugin

File: src/main/java/mate/academy/spring/config/SecurityConfig.java [hunk 2]
```diff
@@ -1,27 +1,50 @@
...
+        auth.userDetailsService(userDetailsService).passwordEncoder(passwordEncoder);
     }
 
     protected void configure(HttpSecurity http) throws Exception {
         http
                 .authorizeRequests()
+                .antMatchers("/register", "/inject").permitAll()
+                .antMatchers(HttpMethod.GET, "/cinema-halls",
+                        "/movies", "/movie-sessions/available",
+                        "/movie-sessions/**").hasAnyRole(USER_ROLE, ADMIN_ROLE)
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/config/SecurityConfig.java:36]:
```suggestion
                        "/movies",
                        "/movie-sessions/**").hasAnyRole(USER_ROLE, ADMIN_ROLE)
```

File: src/main/java/mate/academy/spring/controller/UserController.java [hunk 3]
```diff
@@ -27,4 +29,36 @@ public UserResponseDto findByEmail(@RequestParam String email) {
                 () -> new RuntimeException("User with email " + email + " not found"));
         return userResponseDtoMapper.mapToDto(user);
     }
+
+    @GetMapping("/")
+    public List<UserResponseDto> getAll() {
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/controller/UserController.java:34]: let's add this endpoint into SecurityConfig (I believe only admin should be able to view all users)

File: src/main/java/mate/academy/spring/controller/UserController.java [hunk 4]
```diff
@@ -27,4 +29,36 @@ public UserResponseDto findByEmail(@RequestParam String email) {
...
+        david.setPassword("Jonson");
+
+        User mike = new User();
+        mike.setEmail("Mike@gm.com");
+        mike.setPassword("Shoolermaster");
+
+        userService.add(bob);
+        userService.add(alice);
+        userService.add(david);
+        userService.add(mike);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/controller/UserController.java:61]: let's maybe register them instead?

File: src/main/java/mate/academy/spring/dao/RoleDao.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.spring.dao;
+
+import java.util.Optional;
+import mate.academy.spring.model.Role;
+
+public interface RoleDao {
+
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/dao/RoleDao.java:7]: redundant empty line

File: src/main/java/mate/academy/spring/exception/DataProcessingException.java [hunk 6]
```diff
@@ -1,6 +1,11 @@
 package mate.academy.spring.exception;
 
 public class DataProcessingException extends RuntimeException {
+
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/exception/DataProcessingException.java:4]: redundant empty line

File: src/main/java/mate/academy/spring/security/CustomUserDetailsService.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
+    @Autowired
+    private UserService userService;
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/security/CustomUserDetailsService.java:16]: please better use autowiring through constructor

File: src/main/java/mate/academy/spring/service/mapper/UserMapper.java [hunk 8]
```diff
@@ -1,16 +1,27 @@
...
     @Override
     public UserResponseDto mapToDto(User user) {
         UserResponseDto responseDto = new UserResponseDto();
         responseDto.setId(user.getId());
         responseDto.setEmail(user.getEmail());
         return responseDto;
     }
+
+    @Override
+    public User mapToModel(UserRequestDto dto) {
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/service/mapper/UserMapper.java:21]: where do you use this method?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (oleh-mazurkov) [reply pom.xml:134]: Yes... deleted repeated plugin

File: src/main/java/mate/academy/spring/config/SecurityConfig.java (hunk 2 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/config/SecurityConfig.java:36]: Ok

File: src/main/java/mate/academy/spring/controller/UserController.java (hunk 3 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/controller/UserController.java:34]: added - row 40 in SC

File: src/main/java/mate/academy/spring/controller/UserController.java (hunk 4 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/controller/UserController.java:61]: Ok. I create class InjectController for first enter in DB with Alice and Bob... But I can`t understand how to make the first login (base of User is empty - so I can`t enter in DB))

File: src/main/java/mate/academy/spring/dao/RoleDao.java (hunk 5 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/dao/RoleDao.java:7]: Ok

File: src/main/java/mate/academy/spring/security/CustomUserDetailsService.java (hunk 7 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/security/CustomUserDetailsService.java:16]: Ok

File: src/main/java/mate/academy/spring/service/mapper/UserMapper.java (hunk 8 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/service/mapper/UserMapper.java:21]: Yes, now nowhere)) (it was in processing "my modified" of UserDao... but my additional methods were in AbstructDao))

File: src/main/java/mate/academy/spring/controller/UserController.java (hunk 4 above)
Comment (KseniiaMakarova) [reply src/main/java/mate/academy/spring/controller/UserController.java]: https://www.toolsqa.com/postman/basic-authentication-in-postman/
check this article with screenshots

File: src/main/java/mate/academy/spring/controller/InjectController.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+@RequestMapping("/inject")
+public class InjectController {
+    @Autowired
+    private RoleService roleService;
+
+    @Autowired
+    private AuthenticationService authenticationService;
+
+    @Autowired
+    private UserService userService;
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/controller/InjectController.java:22]: please use annotation through constructor

File: src/main/java/mate/academy/spring/controller/InjectController.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    @GetMapping
+    public String injectData() {
+        roleService.add(new Role(Role.RoleName.ADMIN));
+        roleService.add(new Role(Role.RoleName.USER));
+        authenticationService.register("bob@i.ua","1234");
+        User alice = new User();
+        alice.setEmail("alice@i.ua");
+        alice.setPassword("1234");
+        alice.setRoles(Set.of(roleService.getByName("ADMIN")));
+        userService.add(alice);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/controller/InjectController.java:33]: can we register this user as well?

File: src/main/java/mate/academy/spring/dao/impl/RoleDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    public RoleDaoImpl(SessionFactory factory) {
+        super(factory, Role.class);
+    }
+
+    @Override
+    public Optional<Role> getByName(String roleName) {
+        try (Session session = factory.openSession()) {
+            Query<Role> getByName = session.createQuery(
+                    "FROM Role WHERE rolename =: rolename", Role.class);
+            getByName.setParameter("rolename", roleName);
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/dao/impl/RoleDaoImpl.java:24]: when the field is enum now (now String - does it work?

File: src/main/java/mate/academy/spring/security/CustomUserDetailsService.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        Optional<User> userOptional = userService.findByEmail(email);
+        if (userOptional.isEmpty()) {
+            throw new UsernameNotFoundException("User not found");
+        }
+        User user = userOptional.get();
+        org.springframework.security.core.userdetails.User.UserBuilder builder =
+                org.springframework.security.core.userdetails.User.withUsername(email);
+        builder.password(user.getPassword());
+        builder.authorities(user.getRoles().stream().map(Role::getRoleName).toArray(String[]::new));
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/security/CustomUserDetailsService.java:32]: getRoleName() returns RoleName, not String?

File: src/main/java/mate/academy/spring/service/impl/AuthenticationServiceImpl.java [hunk 13]
```diff
@@ -10,18 +12,22 @@
...
         this.shoppingCartService = shoppingCartService;
+        this.roleService = roleService;
     }
 
     @Override
     public User register(String email, String password) {
         User user = new User();
         user.setEmail(email);
         user.setPassword(password);
+        user.setRoles(Set.of(roleService.getByName("USER")));
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/service/impl/AuthenticationServiceImpl.java:30]: let's use your enum here

File: src/main/java/mate/academy/spring/service/mapper/UserMapper.java [hunk 14]
```diff
@@ -6,6 +6,7 @@
 
 @Component
 public class UserMapper implements ResponseDtoMapper<UserResponseDto, User> {
+
```
Comment (KseniiaMakarova) [inline src/main/java/mate/academy/spring/service/mapper/UserMapper.java:9]: redundant empty line

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/spring/controller/InjectController.java (hunk 9 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/controller/InjectController.java:22]: Ok

File: src/main/java/mate/academy/spring/controller/InjectController.java (hunk 10 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/controller/InjectController.java:33]: ))such variant was in the lesson)) But thank's - deleting this help to solve my problem))

File: src/main/java/mate/academy/spring/dao/impl/RoleDaoImpl.java (hunk 11 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/dao/impl/RoleDaoImpl.java:24]: Changed

File: src/main/java/mate/academy/spring/security/CustomUserDetailsService.java (hunk 12 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/security/CustomUserDetailsService.java:32]: changed

File: src/main/java/mate/academy/spring/service/mapper/UserMapper.java (hunk 14 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/service/mapper/UserMapper.java:9]: Ok

File: src/main/java/mate/academy/spring/config/SecurityConfig.java [hunk 15]
```diff
@@ -1,27 +1,50 @@
...
     protected void configure(HttpSecurity http) throws Exception {
         http
                 .authorizeRequests()
+                .antMatchers("/register", "/inject").permitAll()
+                .antMatchers(HttpMethod.GET, "/cinema-halls",
+                        "/movies",
+                        "/movie-sessions/**").hasAnyRole(USER_ROLE, ADMIN_ROLE)
+                .antMatchers(HttpMethod.POST, "/movies",
+                        "/movie-sessions", "/cinema-halls").hasRole(ADMIN_ROLE)
+
```
Comment (MaliukDaria) [inline src/main/java/mate/academy/spring/config/SecurityConfig.java:39]: redundant empty line

File: src/main/java/mate/academy/spring/controller/InjectController.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        this.authenticationService = authenticationService;
+        this.userService = userService;
+    }
+
+    @GetMapping
+    public String injectData() {
+        roleService.add(new Role(Role.RoleName.USER));
+        roleService.add(new Role(Role.RoleName.ADMIN));
+        authenticationService.register("alice@i.ua","1234");
+        authenticationService.register("bob@i.ua","1234");
```
Comment (MaliukDaria) [inline src/main/java/mate/academy/spring/controller/InjectController.java:32]: I think we should redister some `admin` here

File: src/main/java/mate/academy/spring/security/CustomUserDetailsService.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+        this.userService = userService;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        Optional<User> userOptional = userService.findByEmail(email);
+
+        if (userOptional.isEmpty()) {
+            throw new UsernameNotFoundException("User not found");
+        }
```
Comment (MaliukDaria) [inline src/main/java/mate/academy/spring/security/CustomUserDetailsService.java:28]: do we need this?

File: pom.xml [hunk 18]
```diff
@@ -24,17 +24,17 @@
...
             <groupId>org.springframework</groupId>
             <artifactId>spring-orm</artifactId>
-            <version>${spring.version}</version>
+            <version>5.3.16</version>
         </dependency>
         <dependency>
             <groupId>org.springframework</groupId>
             <artifactId>spring-webmvc</artifactId>
-            <version>${spring.version}</version>
+            <version>5.3.16</version>
```
Comment (MaliukDaria) [inline pom.xml:37]: I think we have to use one spring version in all spring dependencies

File: src/main/java/mate/academy/spring/controller/InjectController.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/inject")
+public class InjectController {
+    private RoleService roleService;
+    private AuthenticationService authenticationService;
+    private UserService userService;
+
+    @Autowired
```
Comment (MaliukDaria) [inline src/main/java/mate/academy/spring/controller/InjectController.java:19]: we have to use  @Autowired annotation everywhere above constructors or nowhere(for code consistency)

File: src/main/java/mate/academy/spring/controller/UserController.java [hunk 20]
```diff
@@ -27,4 +29,11 @@ public UserResponseDto findByEmail(@RequestParam String email) {
                 () -> new RuntimeException("User with email " + email + " not found"));
         return userResponseDtoMapper.mapToDto(user);
     }
+
+    @GetMapping("/")
```
Comment (MaliukDaria) [inline src/main/java/mate/academy/spring/controller/UserController.java:33]:
```suggestion
    @GetMapping
```

**Review by MaliukDaria — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/spring/controller/InjectController.java (hunk 16 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/controller/InjectController.java:32]: Ok

File: src/main/java/mate/academy/spring/config/SecurityConfig.java (hunk 15 above)
Comment (oleh-mazurkov) [reply src/main/java/mate/academy/spring/config/SecurityConfig.java:39]: Ok

**Pull request closed without merging**